File: tree/1372_longest_zigzag_bst.py
# Nathan Zhu May 7th, 2020. Saw Amber today while walking with Renying, they are abouta have finals.
# Leetcode 1372 | medium | not too bad
# Category: tree
# My runtime on this one is N^2, but we should be able to get O(N)

def longestZigZag(root):
    """
    :type root: TreeNode
    :rtype: int
    """
    ret = [0]
    
    def helper(root, lastdir, n):
        ret[0] = max(ret[0], n)
        if not root: return
        
        if lastdir == "left":
            helper(root.right, "right", n + 1)
            helper(root.left, "left", 0)
        else:
            helper(root.left, "left", n + 1)
            helper(root.right, "right", 0)
            
        
    helper(root.left, "left", 0)
    helper(root.right, "right", 0)
    return ret[0]

File: tree/1372_longest_zigzag_bst.cpp
/* Nathan Zhu Friday April 9th, 2020.  Think I am getting better at C++ now, 
*  Leetcode 1372 | medium | medium
*  Category: Binary tree
*
*  This is O(N) instead of the N^2 soln I did before
*/

#include <tuple>

using namespace std;
class TreeNode{
    public:
    TreeNode* left;
    TreeNode* right;
};

tuple<int, int, int> helper(TreeNode* root){
    if(!root) return {0, 0, 0};
    
    auto [ll, lr, totl] = helper(root->left);
    auto [rl, rr, totr] = helper(root->right);
    int ret = max(max(lr + 1, rl + 1), max(totl, totr));
    
    return {lr + 1, rl + 1, ret};

}
int longestZigZag(TreeNode* root) {
    return get<2>(helper(root)) - 1;
}